### PR TITLE
Add BYOK support for ehsm

### PR DESCRIPTION
**Pull request by ForLisiteng**
1. new interface: GetParametersForImport and ImportKeyMaterial.
2. complete BYOK function in python/rust/go sdk and add test

    test: pass unit test

### CODE REVIEW

File: core/App/ffi_operation.cpp [hunk 1]
```diff
@@ -275,7 +275,7 @@ extern "C"
             goto out;
         }
 
-        if (master_key_tmp.keybloblen == 0 || master_key_tmp.keybloblen > UINT16_MAX)
+        if ((master_key_tmp.keybloblen == 0 && master_key_tmp.metadata.origin != EH_EXTERNAL_KEY) || master_key_tmp.keybloblen > UINT16_MAX)
         {
```
Comment (syan10) [inline core/App/ffi_operation.cpp:279]: all the key need to set as EH_EXTERNAL_KEY?

File: core/App/ffi_operation.cpp [hunk 2]
```diff
@@ -390,6 +390,298 @@ extern "C"
...
+        }
+
+        hmac_str = base64_encode(hmac->data, hmac->datalen);
+        if (hmac_str != signature)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+        if (timestamp > time_token + 86400000)
```
Comment (syan10) [inline core/App/ffi_operation.cpp:489]: 86400000 means?

File: core/App/ffi_operation.cpp [hunk 3]
```diff
@@ -390,6 +390,298 @@ extern "C"
...
+        hmac->datalen = EH_HMAC_SHA256_SIZE;
+
+        if (sessionkeyblob == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("SessionKeyBlob is empty.");
+            goto out;
+        }
+
+        ret = GenerateTokenHmac(sessionkeyblob, importToken, hmac);
```
Comment (syan10) [inline core/App/ffi_operation.cpp:474]: sessionkeyblob -> sessionkey

File: core/App/ffi_operation.cpp [hunk 4]
```diff
@@ -390,6 +390,298 @@ extern "C"
...
+
+        // generate a 32B cmk.
+        sessioncmk = (ehsm_keyblob_t *)malloc(APPEND_SIZE_TO_KEYBLOB_T(80));
+
+        sessioncmk->metadata.keyspec = EH_AES_GCM_256;
+        sessioncmk->metadata.origin = EH_INTERNAL_KEY;
+        sessioncmk->metadata.keyusage = EH_KEYUSAGE_ENCRYPT_DECRYPT;
+        sessioncmk->keybloblen = 80;
+
+        ret = CreateKey(sessioncmk);
```
Comment (syan10) [inline core/App/ffi_operation.cpp:648]: what's  difference about the sessioncmk and sessionkeyblob in above function? please unify the name

File: core/App/ffi_operation.cpp (hunk 1 above)
Comment (ForLisiteng) [reply core/App/ffi_operation.cpp:279]: yes,only keybloblen of external key is 0.

File: core/App/ffi_operation.cpp (hunk 2 above)
Comment (ForLisiteng) [reply core/App/ffi_operation.cpp:489]: 24 hours, is there a better way to express it, as a macro?

File: core/Enclave/enclave_hsm.cpp [hunk 5]
```diff
@@ -202,6 +235,227 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
...
+    int paddingMode = getPaddingMode(padding_mode);
+    if (paddingMode != RSA_PKCS1_PADDING && paddingMode != RSA_PKCS1_OAEP_PADDING)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    uint8_t *rsa_keypair = NULL;
+    BIO *bio = NULL;
+    RSA *rsa_prikey = NULL;
+
+    uint32_t keylength = ehsm_get_gcm_ciphertext_size((sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    switch (keylength)
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:384]: what length is expected? 
need to check it's not UINT32_MAX

File: core/Enclave/enclave_hsm.cpp [hunk 6]
```diff
@@ -202,6 +235,227 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
...
+    {
+    case 0:
+    case 16:
+    case 24:
+    case 32:
+        ret = SGX_ERROR_INVALID_STATE;
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:390]: no need to return blob error for cmk now.

File: test/test_kms_with_cli.py [hunk 7]
```diff
@@ -62,6 +63,33 @@ def test_asymmetricKey_encrypt_decrypt(base_url, headers):
...
     print('====================test_asymmetricKey_encrypt_decrypt end===========================')
 
+def test_import_key(base_url, headers):
+    print('====================test_import_key start===========================')
+    key_GCM_256 = createkey.createkey(base_url, "EH_AES_GCM_256", "EH_EXTERNAL_KEY", "EH_KEYUSAGE_ENCRYPT_DECRYPT")
+    symmetricKey = [key_GCM_256]
+
+    aad = str(base64.b64encode("test".encode("utf-8")),'utf-8')
+
+    key = secrets.token_hex(16)
```
Comment (yang8621) [inline test/test_kms_with_cli.py:73]: used ?

File: test/test_kms_with_cli.py [hunk 8]
```diff
@@ -62,6 +63,33 @@ def test_asymmetricKey_encrypt_decrypt(base_url, headers):
...
+
+    for i in symmetricKey:
```
Comment (yang8621) [inline test/test_kms_with_cli.py:75]: need to test 3072 and all paddings

File: test/test_kms_with_cli.py [hunk 9]
```diff
@@ -62,6 +63,33 @@ def test_asymmetricKey_encrypt_decrypt(base_url, headers):
...
+        key_material = asymmetric_encrypt.asymmetric_encrypt(base_url, i, str(base64.b64encode(key.encode("utf-8")), 'utf-8'), "EH_RSA_PKCS1_OAEP")
+        
+        import_key_material.import_key_material(base_url, i, "EH_RSA_PKCS1_OAEP", key_material, importToken)
+
+         # test Encrypt("gcm128")
+        data = str(base64.b64encode("symmetricKeytest".encode("utf-8")),'utf-8')
+        ciphertext = encrypt.encrypt(base_url, i, data, aad)
+
+        # test Decrypt(ciphertext)
+        decrypt.decrypt(base_url, i, ciphertext, aad)
```
Comment (yang8621) [inline test/test_kms_with_cli.py:87]: need compare with plaintext?

File: core/Enclave/enclave_hsm.cpp [hunk 10]
```diff
@@ -347,6 +601,7 @@ sgx_status_t enclave_asymmetric_encrypt(const ehsm_keyblob_t *cmk, size_t cmk_si
 {
     sgx_status_t ret = SGX_ERROR_UNEXPECTED;
 
+#if DEBUG_ONLY
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:604]: why?

File: core/Enclave/enclave_hsm.cpp [hunk 11]
```diff
@@ -376,7 +632,12 @@ sgx_status_t enclave_asymmetric_encrypt(const ehsm_keyblob_t *cmk, size_t cmk_si
         ret = ehsm_sm2_encrypt(cmk, plaintext, ciphertext);
         break;
     default:
+#if DEBUG_ONLY
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:635]: need ?

File: core/Enclave/enclave_hsm.cpp [hunk 12]
```diff
@@ -202,6 +235,227 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
...
+                                 paddingMode);
+    if (retval <= 0)
+    {
+        log_e("failed to make rsa decrypt\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+    import_key->datalen = retval;
+
+    ret = check_aeskey_length(retval, cmk->metadata.keyspec);
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:438]: only aes key supported ?

File: core/Enclave/enclave_hsm.cpp [hunk 13]
```diff
@@ -202,6 +235,227 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
...
+    if (ret != SGX_SUCCESS)
+        goto out;
+
+    memset_s(cmk->keyblob, cmk->keybloblen, 0, cmk->keybloblen);
+
+    ret = ehsm_create_keyblob(import_key->data,
+                              import_key->datalen,
+                              (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    cmk->keybloblen = import_key->datalen + sizeof(sgx_aes_gcm_data_ex_t);
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:448]: Is it possible the memory allocated before is not sufficient to store? 
If we support 64B key, how about here ?

File: core/Enclave/enclave_hsm.cpp [hunk 14]
```diff
@@ -995,11 +1255,57 @@ sgx_status_t enclave_generate_hmac(ehsm_keyblob_t *cmk, uint32_t cmk_size,
...
-    if (rawApiKey) {
-        memset_s(rawApiKey, APPEND_SIZE_TO_DATA_T(rawApiKey->datalen), 0, APPEND_SIZE_TO_DATA_T(rawApiKey->datalen));
-    }
-    // free allocations
+    SAFE_MEMSET(rawApiKey, APPEND_SIZE_TO_DATA_T(rawApiKey->datalen), 0, APPEND_SIZE_TO_DATA_T(rawApiKey->datalen));
     SAFE_FREE(rawApiKey);
     return ret;
+}
+
+sgx_status_t enclave_generate_token_hmac(ehsm_keyblob_t *cmk, uint32_t cmk_size,
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:1263]: cmk mean customer master key. 

Here your cmk refers to the key to generate HMAC signature, right?

File: core/Enclave/enclave_hsm.cpp (hunk 6 above)
Comment (ForLisiteng) [reply core/Enclave/enclave_hsm.cpp]: done

File: core/Enclave/enclave_hsm.cpp (hunk 5 above)
Comment (ForLisiteng) [reply core/Enclave/enclave_hsm.cpp]: I just deleted it

File: test/test_kms_with_cli.py (hunk 7 above)
Comment (ForLisiteng) [reply test/test_kms_with_cli.py]: yes,the key is used to be encrypted by the rsa algorithm.

File: test/test_kms_with_cli.py (hunk 8 above)
Comment (ForLisiteng) [reply test/test_kms_with_cli.py:83]: fix

File: core/Enclave/enclave_hsm.cpp (hunk 12 above)
Comment (ForLisiteng) [reply core/Enclave/enclave_hsm.cpp]: added sm4.

File: core/Enclave/enclave_hsm.cpp (hunk 14 above)
Comment (ForLisiteng) [reply core/Enclave/enclave_hsm.cpp:1252]: fix

File: test/test_kms_with_cli.py (hunk 9 above)
Comment (ForLisiteng) [reply test/test_kms_with_cli.py]: other cases is not compare

File: core/App/ffi_operation.cpp (hunk 4 above)
Comment (ForLisiteng) [reply core/App/ffi_operation.cpp]: fix

File: core/App/ffi_operation.cpp (hunk 3 above)
Comment (ForLisiteng) [reply core/App/ffi_operation.cpp]: fix

File: core/App/ffi_operation.cpp (hunk 1 above)
Comment (ForLisiteng) [reply core/App/ffi_operation.cpp:279]: yes

File: core/App/ehsm_provider.cpp [hunk 15]
```diff
@@ -296,6 +296,12 @@ uint32_t EHSM_FFI_CALL(const char *reqJson, char *respJson)
     case EH_GEN_HMAC:
         ffi_generateHmac(payloadJson, respJson);
```
Comment (syan10) [inline core/App/ehsm_provider.cpp:297]: 整个patch太大了， 拆成几个小的commit：
1. add support for external key origin
2. import key相关的api的实现及unit test
3. go 相关的改动
4. rust相关的改动

File: core/App/ehsm_provider.cpp [hunk 16]
```diff
@@ -493,6 +499,55 @@ ehsm_status_t GetPublicKey(ehsm_keyblob_t *cmk,
         return EH_OK;
 }
 
+ehsm_status_t ImportKeyMaterial(ehsm_keyblob_t *cmk, ehsm_padding_mode_t padding_mode, ehsm_data_t *key_material)
```
Comment (yang8621) [inline core/App/ehsm_provider.cpp:502]: add static for internal functions

File: core/Enclave/enclave_hsm.cpp [hunk 17]
```diff
@@ -202,6 +237,214 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
...
+    bio = BIO_new_mem_buf(rsa_keypair, -1); // use -1 to auto compute length
+    if (bio == NULL)
+    {
+        log_e("failed to load key pem\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    PEM_read_bio_RSAPrivateKey(bio, &rsa_prikey, NULL, NULL);
+    if (rsa_prikey == NULL)
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:404]: is rsa_prikey freed and cleared?

File: core/Enclave/enclave_hsm.cpp [hunk 18]
```diff
@@ -202,6 +237,214 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
...
+    }
+
+    import_key = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(RSA_size(rsa_prikey)));
+    memset(import_key, 0, APPEND_SIZE_TO_DATA_T(RSA_size(rsa_prikey)));
+
+    retval = RSA_private_decrypt(key_material->datalen,
+                                 key_material->data,
+                                 import_key->data,
+                                 rsa_prikey,
+                                 paddingMode);
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:418]: can reuse existing ehsm_rsa_decrypt() ?

File: core/Enclave/enclave_hsm.cpp [hunk 19]
```diff
@@ -97,6 +101,31 @@ static size_t get_signature_length(ehsm_keyspec_t keyspec)
     }
 }
 
+static sgx_status_t check_key_length(int retval, ehsm_keyspec_t keyspec)
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:104]: retval -> keylen

File: core/App/ffi_operation.cpp [hunk 20]
```diff
@@ -390,6 +390,283 @@ extern "C"
...
+        ret = GenerateTokenHmac(sessionkey, importToken, hmac);
+        if (ret != EH_OK)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Generate tokenHmac failed.");
+            goto out;
+        }
+
+        hmac_str = base64_encode(hmac->data, hmac->datalen);
+        if (strcmp(hmac_str.c_str(), signature.c_str()))
```
Comment (yang8621) [inline core/App/ffi_operation.cpp:482]: do not use strcmp

use something like this:
int consttime_memcmp(const void *in_a, const void *in_b, size_t len)
{
  size_t i;
  const unsigned char *a = in_a;
  const unsigned char *b = in_b;
  unsigned char x = 0;
    for (i = 0; i < len; i++)
  	x |= a[i] ^ b[i];
    return x;
  }


Or use CRYPTO_memcmp in openssl.

File: core/App/ehsm_provider.cpp [hunk 21]
```diff
@@ -493,6 +499,55 @@ ehsm_status_t GetPublicKey(ehsm_keyblob_t *cmk,
...
+                                      key_material,
+                                      APPEND_SIZE_TO_DATA_T(key_material->datalen));
+
+    if (ret != SGX_SUCCESS || sgxStatus != SGX_SUCCESS)
+        return EH_FUNCTION_FAILED;
+    else
+        return EH_OK;
+}
+
+ehsm_status_t GetParametersForImport(ehsm_keyblob_t *cmk, ehsm_keyspec_t keyspec, ehsm_data_t *pubkey)
```
Comment (yang8621) [inline core/App/ehsm_provider.cpp:526]: static

File: core/App/ffi_operation.cpp [hunk 22]
```diff
@@ -390,6 +390,283 @@ extern "C"
...
+        if (hmac == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+        hmac->datalen = EH_HMAC_SHA256_SIZE;
+
+        // generate a 32B cmk.
+        sessionkey = (ehsm_keyblob_t *)malloc(APPEND_SIZE_TO_KEYBLOB_T(80));
```
Comment (yang8621) [inline core/App/ffi_operation.cpp:625]: where does 80 come from?

File: core/Enclave/enclave_hsm.cpp [hunk 23]
```diff
@@ -202,6 +237,214 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
...
+
+    bio = BIO_new(BIO_s_mem());
+    if (bio == NULL)
+        goto out;
+
+    if (!PEM_write_bio_RSAPublicKey(bio, rsa_keypair))
+        goto out;
+
+    // pend pubkey
+    if (pubkey->datalen == 0)
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:312]: can reuse sth in ehsm_calc_keyblob_size() ?

File: ehsm_kms_service/ehsm_kms_server.js [hunk 24]
```diff
@@ -63,6 +64,8 @@ const server = (DB) => {
      */
     const { timer: nonce_cache_timer, nonce_database } = _nonce_cache_timer()
 
+    _token_time_verify(DB)
```
Comment (yang8621) [inline ehsm_kms_service/ehsm_kms_server.js:67]: will be called for all requests? 
should be called only for import key material?

File: ehsm_kms_service/function.js [hunk 25]
```diff
@@ -81,6 +81,57 @@ const _nonce_cache_timer = () => {
...
+                            })
+                    }
+                })
+                .catch((err) => {
+                    logger.error(err)
+                })
+        } catch (err) {
+            logger.error(err)
+        }
+    }, Definition.TOKEN_LOOP_CLEAR_TIME)
```
Comment (yang8621) [inline ehsm_kms_service/function.js:132]: if is expired, can the error return to the user?

File: ehsm_kms_service/payload_checker.js [hunk 26]
```diff
@@ -151,6 +151,24 @@ const cryptographic_params = {
...
   },
+  [KMS_ACTION.cryptographic.GetParametersForImport]: {
+    keyid,
+    keyspec: {
+      type: PARAM_DATA_TYPE.CONST,
+      arr: Object.keys(ehsm_keySpec_t),
+      required: true,
+    },
+  },
+  [KMS_ACTION.cryptographic.ImportKeyMaterial]: {
```
Comment (yang8621) [inline ehsm_kms_service/payload_checker.js:162]: verify timestamp is not expired.

Comment (yang8621) [reply ehsm_kms_service/payload_checker.js:162]: test 1. expired token, 2. HMAC wrong (change any byte in token).

File: ehsm_kms_service/constant.js [hunk 27]
```diff
@@ -12,8 +12,9 @@ const Definition = {
     SM_SECRET_VERSION_STAGE_CURRENT: 1,
     SM_SECRET_VERSION_STAGE_PREVIOUS: 0,
     DEFAULT_DELETE_RECOVERY_DAYS: 30,
-    MAX_NONCE_LEN: 64, 
-    FFI_BUFFER_SIZE: 10000
+    MAX_NONCE_LEN: 64,
+    FFI_BUFFER_SIZE: 10000,
+    TOKEN_LOOP_CLEAR_TIME: 60 * 60 * 1000
```
Comment (yang8621) [inline ehsm_kms_service/constant.js:17]: 12 *

File: core/App/ffi_operation.cpp (hunk 22 above)
Comment (ForLisiteng) [reply core/App/ffi_operation.cpp:625]: 80: 32 + sizeof(sgx_aes_gcm_data_ex_t)

File: ehsm_kms_service/function.js [hunk 28]
```diff
@@ -81,6 +81,57 @@ const _nonce_cache_timer = () => {
...
+    setInterval(() => {
+        try {
+            const current_time = new Date()
+                .getTime()
+            const query = {
+                selector: {
+                    //Query criteria
+                    token_expired_time: {
+                        $lt: current_time
+                    },
```
Comment (yang8621) [inline ehsm_kms_service/function.js:94]: using keyidA, could keyidB also be selected?

File: core/Enclave/enclave_hsm.cpp (hunk 17 above)
Comment (ForLisiteng) [reply core/Enclave/enclave_hsm.cpp:392]: fix

File: core/Enclave/enclave_hsm.cpp (hunk 18 above)
Comment (ForLisiteng) [reply core/Enclave/enclave_hsm.cpp:406]: Some memory will be wasted.

File: core/Enclave/enclave_hsm.cpp (hunk 19 above)
Comment (ForLisiteng) [reply core/Enclave/enclave_hsm.cpp]: fix

File: core/App/ffi_operation.cpp (hunk 20 above)
Comment (ForLisiteng) [reply core/App/ffi_operation.cpp]: this part is fixed in web.

File: ehsm_kms_service/constant.js (hunk 27 above)
Comment (ForLisiteng) [reply ehsm_kms_service/constant.js]: fix

File: ehsm_kms_service/function.js (hunk 28 above)
Comment (ForLisiteng) [reply ehsm_kms_service/function.js:94]: this function will range all keyid  status that finish GetParametersForImport request, and check time.

File: sdk/go/kms_test.go [hunk 29]
```diff
@@ -186,3 +189,73 @@ func TestInvalidKeyUsage(t *testing.T) {
...
+	msg := "test for encrypt"
+	ciphertext, err := client.Encrypt(keyid, base64.StdEncoding.EncodeToString([]byte(msg)), base64.StdEncoding.EncodeToString([]byte("test")))
+	if err != nil {
+		t.Error(err)
+	}
+	plaintext_b64, err := client.Decrypt(keyid, ciphertext, base64.StdEncoding.EncodeToString([]byte("test")))
+	if err != nil {
+		t.Error(err)
+	}
+	plaintext, _ := base64.StdEncoding.DecodeString(plaintext_b64)
```
Comment (yang8621) [inline sdk/go/kms_test.go:233]: compare plaintext

File: sdk/go/kms_test.go [hunk 30]
```diff
@@ -186,3 +189,73 @@ func TestInvalidKeyUsage(t *testing.T) {
...
+	result, err := client.ImportKeyMaterial(keyid, "EH_RSA_PKCS1_OAEP", base64.StdEncoding.EncodeToString(key_material), importToken)
+	t.Log(result)
+	if err != nil {
+		t.Error(err)
+	}
+	msg := "test for encrypt"
+	ciphertext, err := client.Encrypt(keyid, base64.StdEncoding.EncodeToString([]byte(msg)), base64.StdEncoding.EncodeToString([]byte("test")))
+	if err != nil {
+		t.Error(err)
+	}
```
Comment (yang8621) [inline sdk/go/kms_test.go:228]: encrypt plaintext with key by self, and compare the ciphertext.

File: sdk/go/kms_test.go [hunk 31]
```diff
@@ -186,3 +189,73 @@ func TestInvalidKeyUsage(t *testing.T) {
...
+
+	key := make([]byte, 32)
+	if _, err := rand.Read(key); err != nil {
+		panic(err)
+	}
+
+	key_material, err := rsa.EncryptOAEP(sha1.New(), rand.Reader, pubkey.(*rsa.PublicKey), []byte(key), nil)
```
Comment (yang8621) [inline sdk/go/kms_test.go:213]: no sha1

File: sdk/go/kms_test.go [hunk 32]
```diff
@@ -186,3 +189,73 @@ func TestInvalidKeyUsage(t *testing.T) {
...
+		t.Error(err)
+	}
+
+	result, err := client.ImportKeyMaterial(keyid, "EH_RSA_PKCS1_OAEP", base64.StdEncoding.EncodeToString(key_material), importToken)
+	t.Log(result)
+	if err != nil {
+		t.Error(err)
+	}
+	msg := "test for encrypt"
+	ciphertext, err := client.Encrypt(keyid, base64.StdEncoding.EncodeToString([]byte(msg)), base64.StdEncoding.EncodeToString([]byte("test")))
```
Comment (yang8621) [inline sdk/go/kms_test.go:225]: aad -> NULL

File: docs/API_Reference.md [hunk 33]
```diff
@@ -688,6 +690,154 @@ ehsm-core enclave will decrypt user-supplied ciphertextblob with specified CMK t
...
+  ```python
+    Response= {
+      "code": 200,
+      "message": "success!"
+    }
+  ```
+  *(return to the [Cryptographic Functionalities APIs](#eHSM-REST-API-Reference).)*
+---
+
+## GetPublicKey
```
Comment (yang8621) [inline docs/API_Reference.md:799]: GetParametersForImport ?

File: ehsm_kms_service/function.js [hunk 34]
```diff
@@ -664,6 +723,55 @@ const gen_hmac = async (DB, appid, sign_params) => {
...
+    }
+}
+
+/**
+ * Compare string size on const time.
+ * @param {string} str1 
+ * @param {string} str2 
+ * @returns true | false
+ */
+function consttime_compareStrings(str1, str2) {
```
Comment (yang8621) [inline ehsm_kms_service/function.js:761]: 返回值有点反直觉
consttime_compareStrings -> consttime_equal_compare
{
return true 表示相等吧
}

File: ehsm_kms_service/router.js [hunk 35]
```diff
@@ -162,6 +166,108 @@ const router = async (p) => {
...
+            'keyspec',
+            'origin',
+            'keyState',
+            'sessionkeyBlob',
+          ],
+          limit: 1,
+        }
+        let query_result = await DB.partitionedFind('cmk', query)
+        const { sessionkeyBlob } = query_result.docs[0]
+        if (timestamp_now * 1 > timestamp * 1 + 24 * 60 * 60 * 1000) {
```
Comment (yang8621) [inline ehsm_kms_service/router.js:245]: timestamp 是从172行拿到的？

File: ehsm_kms_service/router.js [hunk 36]
```diff
@@ -162,6 +166,108 @@ const router = async (p) => {
...
+      } catch (error) {
+        logger.error(error)
+        res.send(_result(500, 'GetParametersForImport failed.'))
+      }
+      break
+    case KMS_ACTION.cryptographic.ImportKeyMaterial:
+      try {
+        let { keyid, padding_mode, key_material, importToken } = payload
+        const timestamp_now = req.body['timestamp']
+        const { keyid: keyid_token, timestamp, hmac: signature } = await JSON.parse(base64_decode(importToken))
```
Comment (yang8621) [inline ehsm_kms_service/router.js:220]: 如果importToken 里只有2项，或者不是json， 会挂掉吗 ？ 
还是说这三个值是undefined？

File: ehsm_kms_service/router.js [hunk 37]
```diff
@@ -162,6 +166,108 @@ const router = async (p) => {
...
+            'keyspec',
+            'origin',
+            'keyState',
+            'sessionkeyBlob',
+            'token_expired_time',
+          ],
+          limit: 1,
+        }
+        let cmks_res = await DB.partitionedFind('cmk', query)
+        cmks_res.docs[0].token_expired_time = timestamp * 1 + 24 * 60 * 60 * 1000
```
Comment (yang8621) [inline ehsm_kms_service/router.js:202]: 24 * 60 * 60 * 1000  define a macro

File: ehsm_kms_service/router.js (hunk 35 above)
Comment (yang8621) [inline ehsm_kms_service/router.js:245]: 24 * 60 * 60 * 1000  define a macro

File: ehsm_kms_service/router.js [hunk 38]
```diff
@@ -162,6 +166,108 @@ const router = async (p) => {
         napi_res && res.send(napi_res)
       } catch (error) { }
       break
+    case KMS_ACTION.cryptographic.GetParametersForImport:
+      try {
+        let { keyid, keyspec } = payload
+        let timestamp = req.body['timestamp']
```
Comment (yang8621) [inline ehsm_kms_service/router.js:172]: use server timestamp

File: ehsm_kms_service/router.js (hunk 35 above)
Comment (yang8621) [reply ehsm_kms_service/router.js:245]: use server timestamp

File: sdk/go/kms_test.go [hunk 39]
```diff
@@ -186,3 +187,83 @@ func TestInvalidKeyUsage(t *testing.T) {
...
+		for _, warpping_keyspec := range warpping_keyspec {
+			pubkey, importToken, err := client.GetParametersForImport(keyid, warpping_keyspec)
+			if err != nil {
+				t.Error(err)
+			}
+			key, err := generateRandomKey(keyspec)
+			if err != nil {
+				t.Error(err)
+			}
+			if _, err := rand.Read(key); err != nil {
```
Comment (yang8621) [inline sdk/go/kms_test.go:213]: 放到generateRandomKey里面

File: sdk/go/get_parameters_for_import.go [hunk 40]
```diff
@@ -0,0 +1,117 @@
...
+	payload.Set("keyspec", keyspec)
+	params := c.initParams(payload)
+
+	c.modifyLock.Lock()
+	defer c.modifyLock.Unlock()
+
+	// call ehsm kms
+	resp, err := c.doPost(params, "GetParametersForImport")
+	if err != nil {
+		return "", "", err
```
Comment (yang8621) [inline sdk/go/get_parameters_for_import.go:36]: use nil but not ""

File: sdk/rust/src/client.rs [hunk 41]
```diff
@@ -99,6 +105,51 @@ pub fn is_base64(s: &str) -> bool {
 	base64::decode_config(s, base64::STANDARD).is_ok()
 }
 
+fn generate_random_key(keyspec: &str) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
+    match keyspec {
+        "EH_AES_GCM_128" | "EH_SM4_CBC" | "EH_SM4_CTR" => {
+            let mut key = vec![0u8; 16];
+            rand::thread_rng().fill(&mut key[..]);
```
Comment (yang8621) [inline sdk/rust/src/client.rs:112]: rust 不要用 rand::thread_rng()
用ring

**Review by yang8621 — APPROVED** [review]

**Review by yang8621 — APPROVED** [review]

**Review by syan10 — APPROVED** [review]

**Pull request merged**